### ISSUE TITLE
combo driver causes dependency cycle

### ISSUE THREAD

**Issue by justjulez420** [Arminando/ActionPoser#1]
Hi!
i noticed that the drivers on the final modifiers don't always update correctly even tho all the values in the actionposer menu are correct. so i looked at the console and i keep getting this dependency cycle errors from what i assume to be the combo driver type: 

Dependency cycle detected:
  ARArmature/PARAMETERS_EVAL() depends on
  ARArmature/DRIVER(ap_poses[0].influence) via 'Driver -> Driven Property'
  ARArmature/PARAMETERS_EVAL() via 'RNA Target -> Driver'

i double checked and couldn't find anything wrong with my setup. but here is the rig, just in case:
https://drive.google.com/file/d/13hOwor-lYjbWog6fYBt3eAbxaa6S2JaN/view?usp=sharing

**Comment by justjulez420** [comment]
just to be completely sure that its not a problem with my setup, i tried creating a combo driver in a brand new blend file - same result. 
here's the new blend file: https://drive.google.com/file/d/11kVopobEBRK0gpy68PJJU0xBkv8B2Lem/view?usp=sharing

**Comment by arminHalac** [comment]
Hi! I will look into it as soon as I can. Not sure from the cycle error message what the problem is but shouldn't be hard to figure out once I dig into it.
Will get back to you soon :)

**Comment by Arminando** [comment]
Hi @justjulez420 ,
I inspected what was happening and could not find anything that is out of the ordinary. My guess is that the depsgraph is unable to figure out the order in which to evaluate the drivers so it throws out the cycle message. But the rig seems to work regardless of that.
I will continue looking for a workaround but at least it seems like this is not breaking rigs.

**Comment by justjulez420** [comment]
Thanks for taking the time!
Unfortunately, it can cause problems later on. While it works on it's own and generally seems to work when moving the control bones manually, the driver doesn't update properly in playback (or the final render) when used in a heavy scene. the problem seems to get worse the more my pc struggles from what i can tell
![](https://i.imgur.com/brtKUqN.gif)

**Comment by Arminando** [comment]
My guess is that the evaluation is multithreaded, and since I am using the same property for regular poses and combos it is unable to run them in the proper order. Will split them into different props and see if that fixes it. Thanks for the feedback again. It is super useful!

**Comment by justjulez420** [comment]
Any progress so far?

**Comment by arminHalac** [comment]
I would need more time, unfortunately. Tried the approach of splitting them into unique properties to see if that fixed things but it didn't. So I probably have to redo how all of the driving works.

**Comment by justjulez420** [comment]
So, i don't really know much python but i managed to throw together this script: https://github.com/justjulez420/ActionPoserFix
It writes the drivers directly to the constraints, avoiding dependency cycles in most cases and eliminating the update issue. Not the most elegant solution but it did the job for me. Thought i might leave it here in case it's any help

**Comment by arminHalac** [comment]
Hey @justjulez420, I finally worked on this. The fix is exactly what you suggested, namely driving the constraints directly. It is on the "time_for_an_overhaul" branch. Would be great if you could test it and let me know if you find any other issues. I tested on Blender 4.0, not sure if something will break for earlier versions.

**Comment by justjulez420** [comment]
Thanks a lot! Tested it on blender 3.6 and can confirm it works there too.
I found one bug though: combo drivers are not being converted to degrees when using rotational values
![example](https://github.com/Arminando/ActionPoser/assets/69090529/10096368-6150-4d02-85d4-6f0b1aa38a68)
Besides that, everything works perfectly

**Comment by arminHalac** [comment]
Good catch. I think I fixed it now. Let me know if you find other issues.
Thank you for testing and reporting the bug
